File: app/src/main/java/knf/kuma/updater/UpdateActivity.java
package knf.kuma.updater;

import android.animation.Animator;
import android.annotation.SuppressLint;
import android.content.Context;
import android.content.Intent;
import android.graphics.Rect;
import android.graphics.drawable.AnimationDrawable;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.os.Environment;
import android.support.annotation.Nullable;
import android.support.v4.content.FileProvider;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.CardView;
import android.util.Log;
import android.view.View;
import android.view.ViewAnimationUtils;
import android.view.WindowManager;
import android.view.animation.Animation;
import android.view.animation.AnimationUtils;
import android.widget.Button;
import android.widget.ProgressBar;
import android.widget.RelativeLayout;
import android.widget.TextView;

import com.thin.downloadmanager.DownloadRequest;
import com.thin.downloadmanager.DownloadStatusListenerV1;
import com.thin.downloadmanager.ThinDownloadManager;

import java.io.File;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import knf.kuma.R;
import xdroid.toaster.Toaster;

public class UpdateActivity extends AppCompatActivity {

    @BindView(R.id.rel_back)
    RelativeLayout back;
    @BindView(R.id.card)
    CardView cardView;
    @BindView(R.id.progress)
    ProgressBar progressBar;
    @BindView(R.id.progress_text)
    TextView progress_indicator;
    @BindView(R.id.download)
    Button download;

    public static void start(Context context) {
        context.startActivity(new Intent(context, UpdateActivity.class));
    }

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_updater);
        ButterKnife.bind(this);
        getWindow().addFlags(WindowManager.LayoutParams.FLAG_TRANSLUCENT_STATUS |
                WindowManager.LayoutParams.FLAG_TRANSLUCENT_NAVIGATION);
        if (savedInstanceState == null) {
            AnimationDrawable animationDrawable = (AnimationDrawable) back.getBackground();
            animationDrawable.setEnterFadeDuration(2500);
            animationDrawable.setExitFadeDuration(2500);
            animationDrawable.start();
            progressBar.setMax(100);
            progressBar.setIndeterminate(true);
            showCard();
        }
    }

    private void showCard() {
        cardView.post(new Runnable() {
            @Override
            public void run() {
                Rect bounds = new Rect();
                cardView.getDrawingRect(bounds);
                int centerX = bounds.centerX();
                int centerY = bounds.centerY();
                int finalRadius = Math.max(bounds.width(), bounds.height());

                Animator anim = ViewAnimationUtils.createCircularReveal(cardView, centerX, centerY, 0f, finalRadius);
                anim.addListener(new Animator.AnimatorListener() {
                    @Override
                    public void onAnimationStart(Animator animation) {

                    }

                    @Override
                    public void onAnimationEnd(Animator animation) {
                        start();
                    }

                    @Override
                    public void onAnimationCancel(Animator animation) {

                    }

                    @Override
                    public void onAnimationRepeat(Animator animation) {

                    }
                });
                cardView.setVisibility(View.VISIBLE);
                anim.start();
            }
        });
    }

    private void start() {
        File file = getUpdate();
        if (file.exists())
            file.delete();
        new ThinDownloadManager().add(new DownloadRequest(Uri.parse("https://github.com/jordyamc/UKIKU/raw/master/app/release/app-release.apk"))
                .setDestinationURI(Uri.fromFile(file))
                .setDownloadResumable(false)
                .setStatusListener(new DownloadStatusListenerV1() {
                    @Override
                    public void onDownloadComplete(DownloadRequest downloadRequest) {
                        prepareForIntall();
                    }

                    @Override
                    public void onDownloadFailed(DownloadRequest downloadRequest, int errorCode, String errorMessage) {
                        Log.e("Update Error", "Code: " + errorCode + " Message: " + errorMessage);
                        Toaster.toast("Error al actualizar: "+errorMessage);
                        finish();
                    }

                    @Override
                    public void onProgress(DownloadRequest downloadRequest, long totalBytes, long downloadedBytes, int progress) {
                        setDownProgress(progress);
                    }
                }));
    }

    @OnClick(R.id.download)
    void install(Button button) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
            Intent intent = new Intent(Intent.ACTION_INSTALL_PACKAGE, FileProvider.getUriForFile(this, "knf.kuma.fileprovider", getUpdate()))
                    .addFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_GRANT_READ_URI_PERMISSION | Intent.FLAG_GRANT_WRITE_URI_PERMISSION)
                    .putExtra(Intent.EXTRA_NOT_UNKNOWN_SOURCE, false)
                    .putExtra(Intent.EXTRA_INSTALLER_PACKAGE_NAME, getPackageName());
            startActivity(intent);
        } else {
            Intent intent = new Intent(Intent.ACTION_VIEW)
                    .setDataAndType(Uri.fromFile(getUpdate()), "application/vnd.android.package-archive")
                    .addFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_GRANT_READ_URI_PERMISSION | Intent.FLAG_GRANT_WRITE_URI_PERMISSION);
            startActivity(intent);
        }
        finish();
    }

    private File getUpdate() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N)
            return new File(getFilesDir(), "update.apk");
        else
            return new File(getDownloadsDir(), "update.apk");
    }

    private File getDownloadsDir() {
        return Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_DOWNLOADS);
    }

    private void setDownProgress(final int progress) {
        runOnUiThread(new Runnable() {
            @SuppressLint("SetTextI18n")
            @Override
            public void run() {
                progressBar.setIndeterminate(false);
                progressBar.setProgress(progress);
                progress_indicator.setText(progress + "%");
            }
        });
    }

    private void prepareForIntall() {
        setDownProgress(100);
        final Animation fadein = AnimationUtils.loadAnimation(this, R.anim.fadein);
        fadein.setDuration(1000);
        final Animation fadeout = AnimationUtils.loadAnimation(this, R.anim.fadeout);
        fadeout.setDuration(1000);
        progress_indicator.post(new Runnable() {
            @Override
            public void run() {
                progress_indicator.setVisibility(View.INVISIBLE);
                progress_indicator.startAnimation(fadeout);
            }
        });
        download.post(new Runnable() {
            @Override
            public void run() {
                download.setVisibility(View.VISIBLE);
                download.startAnimation(fadein);
            }
        });
    }
}

File: app/src/main/java/knf/kuma/updater/UpdateActivity.kt
package knf.kuma.updater

import androidx.activity.addCallback
import android.annotation.SuppressLint
import android.content.Context
import android.content.Intent
import android.graphics.drawable.AnimationDrawable
import android.net.Uri
import android.os.Build
import android.os.Bundle
import android.view.View
import android.view.WindowManager
import android.view.animation.AnimationUtils
import androidx.activity.viewModels
import androidx.core.content.FileProvider
import androidx.lifecycle.Observer
import com.afollestad.materialdialogs.MaterialDialog
import knf.kuma.R
import knf.kuma.commons.getUpdateDir
import knf.kuma.commons.safeShow
import knf.kuma.custom.GenericActivity
import knf.kuma.databinding.ActivityUpdaterBinding
import knf.kuma.download.DownloadManagerCentral
import java.io.File
import androidx.core.net.toUri

class UpdateActivity : GenericActivity() {

    private val binding by lazy { ActivityUpdaterBinding.inflate(layoutInflater) }
    private val updaterViewModel: UpdaterViewModel by viewModels()
    private val update: File by lazy { File.createTempFile("update", ".apk", filesDir) }
    private var isUpdateDownloaded = false

    @SuppressLint("SetTextI18n")
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(binding.root)
        window.addFlags(WindowManager.LayoutParams.FLAG_TRANSLUCENT_STATUS or WindowManager.LayoutParams.FLAG_TRANSLUCENT_NAVIGATION)
        binding.download.setOnClickListener { install() }
        binding.retry.setOnClickListener {
            prepareForStart()
            start()
        }
        binding.progress.max = 100
        val animationDrawable = binding.relBack.background as AnimationDrawable
        if (!animationDrawable.isRunning) {
            animationDrawable.setEnterFadeDuration(2500)
            animationDrawable.setExitFadeDuration(2500)
            animationDrawable.start()
        }
        onBackPressedDispatcher.addCallback(this) {
            if (intent.getBooleanExtra("canExit", true)) {
                isEnabled = false
                onBackPressedDispatcher.onBackPressed()
                isEnabled = true
            }
        }
        start()
    }

    private fun start() {
        updaterViewModel.start(update, "https://github.com/jordyamc/UKIKU/raw/master/app/$getUpdateDir/app-$getUpdateDir.apk")
            .observe(this, Observer {
                when (it.first) {
                    UpdaterType.TYPE_IDLE -> {
                        binding.progress.isIndeterminate = true
                    }

                    UpdaterType.TYPE_PROGRESS -> {
                        setDownProgress(it.second as Int)
                    }

                    UpdaterType.TYPE_ERROR -> {
                        prepareForRetry()
                        onShowAlternativeDownload()
                    }

                    UpdaterType.TYPE_COMPLETED -> {
                        isUpdateDownloaded = true
                        binding.progress.progress = 100
                        binding.progressText.text = "100%"
                        prepareForInstall()
                    }
                }
            })
    }

    override fun onStart() {
        super.onStart()
        if (!intent.getBooleanExtra("canExit", true)) {
            MaterialDialog(this).safeShow {
                message(text = "Parece que la versión ${intent.getStringExtra("version")} está disponible, es obligatoria")
                positiveButton(text = "ok")
            }
        }
    }

    private fun install() {
        DownloadManagerCentral.pauseAll()
        val intent = Intent(Intent.ACTION_INSTALL_PACKAGE, FileProvider.getUriForFile(this, "${applicationContext.packageName}.fileprovider", update))
            .addFlags(Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_GRANT_READ_URI_PERMISSION or Intent.FLAG_GRANT_WRITE_URI_PERMISSION)
            .putExtra(Intent.EXTRA_NOT_UNKNOWN_SOURCE, false)
            .putExtra(Intent.EXTRA_INSTALLER_PACKAGE_NAME, packageName)
        startActivity(intent)
    }

    private fun setDownProgress(p: Int) {
        try {
            binding.progress.apply {
                isIndeterminate = false
                progress = p
            }
            binding.progressText.text = String.format("%d%%", p)
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    private fun prepareForInstall() {
        setDownProgress(100)
        val fadein = AnimationUtils.loadAnimation(this, R.anim.fadein)
        fadein.duration = 1000
        val fadeout = AnimationUtils.loadAnimation(this, R.anim.fadeout)
        fadeout.duration = 1000
        binding.progressText.post {
            with(binding.progressText) {
                visibility = View.INVISIBLE
                startAnimation(fadeout)
            }
        }
        binding.download.post {
            with(binding.download) {
                visibility = View.VISIBLE
                startAnimation(fadein)
            }
        }
    }

    private fun prepareForStart() {
        setDownProgress(100)
        val fadein = AnimationUtils.loadAnimation(this, R.anim.fadein)
        fadein.duration = 1000
        val fadeout = AnimationUtils.loadAnimation(this, R.anim.fadeout)
        fadeout.duration = 1000
        binding.retry.post {
            with(binding.retry) {
                visibility = View.INVISIBLE
                startAnimation(fadeout)
            }
        }
        binding.progressText.post {
            with(binding.progressText) {
                visibility = View.VISIBLE
                startAnimation(fadein)
            }
        }
    }

    private fun prepareForRetry() {
        setDownProgress(100)
        val fadein = AnimationUtils.loadAnimation(this, R.anim.fadein)
        fadein.duration = 1000
        val fadeout = AnimationUtils.loadAnimation(this, R.anim.fadeout)
        fadeout.duration = 1000
        binding.progressText.post {
            with(binding.progressText) {
                visibility = View.INVISIBLE
                startAnimation(fadeout)
            }
        }
        binding.retry.post {
            with(binding.retry) {
                visibility = View.VISIBLE
                startAnimation(fadein)
            }
        }
    }

    private fun onShowAlternativeDownload() {
        MaterialDialog(this).safeShow {
            title(text = "¿Error al actualizar?")
            message(text = "Puedes descargar la actualización desde la pagina web oficial!")
            positiveButton(text = "Descargar") {
                startActivity(Intent(Intent.ACTION_VIEW, "https://ukiku.app".toUri()))
            }
        }
    }

    override fun onResume() {
        super.onResume()
        if (isUpdateDownloaded) {
            onShowAlternativeDownload()
        }
    }

    companion object {
        fun start(context: Context, canExit: Boolean, version: String) {
            context.startActivity(Intent(context, UpdateActivity::class.java).apply {
                putExtra("canExit", canExit)
                putExtra("version", version)
            })
        }
    }
}
